controllers/kafkacluster_controller.go

```diff
@@ -147,84 +142,167 @@ func (r *KafkaClusterReconciler) Reconcile(request ctrl.Request) (ctrl.Result, e
 	}
 
 	//Update rolling upgrade last successful state
-	if instance.Status.State == banzaicloudv1beta1.KafkaClusterRollingUpgrading {
+	if instance.Status.State == v1beta1.KafkaClusterRollingUpgrading {
 		if err := k8sutil.UpdateRollingUpgradeState(r.Client, instance, time.Now(), log); err != nil {
 			return requeueWithError(log, err.Error(), err)
 		}
 	}
 
-	if err := k8sutil.UpdateCRStatus(r.Client, instance, banzaicloudv1beta1.KafkaClusterRunning, log); err != nil {
+	if err := k8sutil.UpdateCRStatus(r.Client, instance, v1beta1.KafkaClusterRunning, log); err != nil {
 		return requeueWithError(log, err.Error(), err)
 	}
 
 	return reconciled()
 }
 
-func (r *KafkaClusterReconciler) checkFinalizers(log logr.Logger, cluster *v1beta1.KafkaCluster) (ctrl.Result, error) {
+func (r *KafkaClusterReconciler) checkFinalizers(ctx context.Context, log logr.Logger, cluster *v1beta1.KafkaCluster) (ctrl.Result, error) {
 	log.Info("KafkaCluster is marked for deletion, checking for children")
+
+	// If the main finalizer is gone then we've already finished up
 	if !util.StringSliceContains(cluster.GetFinalizers(), clusterFinalizer) {
 		return reconciled()
 	}
 
 	var err error
 
-	// Delete all kafkatopics associated with us
-	var childTopics banzaicloudv1alpha1.KafkaTopicList
-	if err = r.Client.List(context.TODO(), &childTopics); err != nil {
+	var namespaces corev1.NamespaceList
+	if err := r.Client.List(ctx, &namespaces); err != nil {
+		return requeueWithError(log, "failed to get namespace list", err)
+	}
+
+	// If we haven't deleted all kafkatopics yet, iterate namespaces and delete all kafkatopics
+	// with the matching label.
+	if util.StringSliceContains(cluster.GetFinalizers(), clusterTopicsFinalizer) {
+		log.Info(fmt.Sprintf("Sending delete kafkatopics request to all namespaces for cluster %s/%s", cluster.Namespace, cluster.Name))
+		for _, ns := range namespaces.Items {
+			if err := r.Client.DeleteAllOf(
+				ctx,
+				&v1alpha1.KafkaTopic{},
+				client.InNamespace(ns.Name),
+				client.MatchingLabels{clusterRefLabel: clusterLabelString(cluster)},
+			); err != nil {
+				if client.IgnoreNotFound(err) != nil {
+					return requeueWithError(log, "failed to send delete request for children kafkatopics", err)
+				}
+				log.Info(fmt.Sprintf("No matching kafkatopics in namespace: %s", ns.Name))
+			}
+
+		}
+		if cluster, err = r.removeFinalizer(ctx, cluster, clusterTopicsFinalizer); err != nil {
+			return requeueWithError(log, "failed to remove topics finalizer from kafkacluster", err)
+		}
+	}
+
+	// If any of the topics still exist, it means their finalizer is still running.
+	// Wait to make sure we have fully cleaned up zookeeper. Also if we delete
+	// our kafkausers before all topics are finished cleaning up, we will lose
+	// our controller certificate.
+	log.Info("Ensuring all topics have finished cleaning up")
+	var childTopics v1alpha1.KafkaTopicList
+	if err = r.Client.List(ctx, &childTopics); err != nil {
```

Comment:
If I saw that right, all the KafkaTopics are labeled with some cluster specific label, why don't we use List filters to get back only those which are relevant for us, with that I think we can remove the `if belongsToCluster` part.